services/device-storage/js/app.js

```diff
@@ -12,124 +12,178 @@
   var _deviceStorages = {};
   var _listeners = {};
 
-  var processSWRequest = function(channel, evt) {
-    // We can get:
-    // * methodName
-    // * onchange
-    // * getDeviceStorage
-    // * addEventListener
-    // * removeEventListener
-    // * dispatchEvent
-    // All the operations have a requestId, and the lock operations also include
-    // a deviceStorage id.
-    var remotePortId = evt.data.remotePortId;
-    var request = evt.data.remoteData;
-    var requestOp = request.data;
+  function addEventTargetEvent(channel, request) {
+    var requestOp = request.remoteData.data;
+    var reqId = request.remoteData.id;
+    var deviceStorageId = requestOp.deviceStorageId;
 
-    function observerTemplate(evt) {
+    function listenerTemplate(evt) {
       channel.postMessage({
         remotePortId: remotePortId,
         data: {
-          id: request.id,
-          data: {
-            path: evt.path,
-            reason: evt.reason
-          }
+          id: reqId,
+          event: window.ServiceHelper.cloneObject(evt)
         }
       });
     }
 
-    function listenerTemplate(evt) {
+    _listeners[reqId] = listenerTemplate;
+    _deviceStorages[deviceStorageId].addEventListener(requestOp.type,
+      _listeners[reqId], requestOp.useCapture);
+  }
+
+  function buildDOMCursorAnswer(operation, channel, request) {
+    var remotePortId = request.remotePortId;
+    // Params for the local operation:
+    var opData = request.remoteData.data.params || [];
+    var reqId = request.remoteData.id;
+    var deviceStorageId = request.remoteData.data.deviceStorageId;
+
+    opData.forEach((param, index) => {
+      if (param === null || typeof param === 'undefined') {
+        opData.splice(index, 1);
```

Comment:
First, I dunno how much forEach is going to like the array being modified while it's being processed... and second, I think what you really want to do is remove the tailing undefines... because

[1,undefined, 2] should still be [1,undefined, 2] and no [1,2]

while
[1,2, undefined, undefined] should be [1,2]